### PR TITLE
feat(core): support shift click zooming

### PR DESCRIPTION
**Pull request by hlyang397**
### Updates
- Shift + click in chart could zoom in based on mouse position
- Dispatch ZoomDomain.CHANGE event whenever zoom domain is changed.

### Demo screenshot or recording
shiftClick

### Review checklist (for reviewers only)
- [ ] Demos all features
- [ ] Documented/annotated
- [ ] Matches UI/UX specs
- [ ] Meets the [code style guide](https://github.com/carbon-design-system/carbon-charts/wiki/Code-style-guide)
- [ ] Accessible
- [ ] Mobile first (responsive)
- [ ] RTL support (bidirectional text)
- [ ] Performant (limited bloat)

### CODE REVIEW

**Review by theiliad — CHANGES_REQUESTED** [review]
again, with this PR we agreed to switch to this behaviour:
- Zoom onscroll (scrolling up would zoom in, scrolling down would zoom out)
- Pan on drag
- Zoom on shift + drag like you have in there currently

**Comment by hlyang397** [comment]
> again, with this PR we agreed to switch to this behaviour:
> 
>     * Zoom onscroll (scrolling up would zoom in, scrolling down would zoom out)
> 
>     * Pan on drag
> 
>     * Zoom on shift + drag like you have in there currently

@theiliad 
1. This PR is only about the zoom on shift + click. Why this PR can't be simple without the other implementations? The other behaviors should be in another PRs.
2. I don't think we agree the other behaviors, especially the pan on drag.

**Comment by theiliad** [comment]
> > again, with this PR we agreed to switch to this behaviour:
> > ```
> > * Zoom onscroll (scrolling up would zoom in, scrolling down would zoom out)
> > 
> > * Pan on drag
> > 
> > * Zoom on shift + drag like you have in there currently
> > ```
> 
> @theiliad
> 
> 1. This PR is only about the zoom on shift + click. Why this PR can't be simple without the other implementations? The other behaviors should be in another PRs.
> 2. I don't think we agree the other behaviors, especially the pan on drag.

@hlyang397 Pan on drag is what's been decided by design. We really need to implement that

**Comment by hlyang397** [comment]
> > > again, with this PR we agreed to switch to this behaviour:
> > > ```
> > > * Zoom onscroll (scrolling up would zoom in, scrolling down would zoom out)
> > > 
> > > * Pan on drag
> > > 
> > > * Zoom on shift + drag like you have in there currently
> > > ```
> > 
> > 
> > @theiliad
> > 
> > 1. This PR is only about the zoom on shift + click. Why this PR can't be simple without the other implementations? The other behaviors should be in another PRs.
> > 2. I don't think we agree the other behaviors, especially the pan on drag.
> 
> @hlyang397 Pan on drag is what's been decided by design. We really need to implement that, as well as zoom onscroll

@theiliad Yes, I think we should provide these capabilities. But my question is why these can't be another PRs?

**Comment by theiliad** [comment]
> > > > again, with this PR we agreed to switch to this behaviour:
> > > > ```
> > > > * Zoom onscroll (scrolling up would zoom in, scrolling down would zoom out)
> > > > 
> > > > * Pan on drag
> > > > 
> > > > * Zoom on shift + drag like you have in there currently
> > > > ```
> > > 
> > > 
> > > @theiliad
> > > 
> > > 1. This PR is only about the zoom on shift + click. Why this PR can't be simple without the other implementations? The other behaviors should be in another PRs.
> > > 2. I don't think we agree the other behaviors, especially the pan on drag.
> > 
> > 
> > @hlyang397 Pan on drag is what's been decided by design. We really need to implement that, as well as zoom onscroll
> 
> @theiliad Yes, I think we should provide these capabilities. But my question is why these can't be another PRs?

Sure as long as we know that we're going to add those piece then I'm good with merging this

File: packages/core/src/services/zoom.ts [hunk 1]
```diff
@@ -64,4 +65,19 @@ export class Zoom extends Service {
 			extent(zoomBarData, (d: any) => d[domainIdentifier])
 		);
 	}
+
+	handleDomainChange(newDomain) {
+		this.model.set({ zoomDomain: newDomain }, { animate: false });
+		this.services.events.dispatchEvent(Events.ZoomDomain.CHANGE, {
+			newDomain
+		});
```
Comment (theiliad) [inline packages/core/src/services/zoom.ts:73]: 👍

File: packages/core/src/components/axes/grid-brush.ts [hunk 2]
```diff
@@ -97,87 +103,76 @@ export class ChartBrush extends Component {
...
+					let leftPoint = clickedX - (width * zoomRatio) / 2;
+					if (leftPoint < 0) {
+						leftPoint = 0;
+					}
+					let rightPoint = clickedX + (width * zoomRatio) / 2;
+					if (rightPoint > width) {
+						rightPoint = width;
+					}
+					// handleZoomDomain assumes max range is [0, width]
+					handleZoomDomain(leftPoint, rightPoint);
```
Comment (theiliad) [inline packages/core/src/components/axes/grid-brush.ts:173]: `handleZoomDomain` might not be the best name, what does this function do? It confuses me a bit too

Comment (theiliad) [reply packages/core/src/components/axes/grid-brush.ts:173]: looks like it's updating the zoom domain, maybe `updateZoomDomain`?

**Review by theiliad — CHANGES_REQUESTED** [review]
beyond the requested changes it's good to go

File: packages/core/src/services/zoom.ts (hunk 1 above)
Comment (theiliad) [reply packages/core/src/services/zoom.ts:73]: no change needed here btw

File: packages/core/src/components/axes/grid-brush.ts (hunk 2 above)
Comment (hlyang397) [reply packages/core/src/components/axes/grid-brush.ts]: I am ok with `updateZoomDomain`.  Updated in https://github.com/carbon-design-system/carbon-charts/pull/773/commits/fb1d4ab541e920409606016791cc45cbe2872896.

**Review by theiliad — APPROVED** [review]

**Review by cal-smith — APPROVED** [review]
Makes sense to me!

**Pull request merged**